### PR TITLE
Add Target-Label to jar manifest without replacing the rest of the jar manifest entries

### PR DESCRIPTION
**Pull request by cheister**
Addressing https://github.com/bazelbuild/rules_jvm_external/issues/292 so that jar manifest entries are preserved when the `Target-Label` manifest entry is added to a stamped jar.

### CODE REVIEW

**Comment by jin** [comment]
Hmm, I wonder if there's a singlejar mode to preserve the existing content? Also the failing test suggests permission issues on Windows.

**Comment by cheister** [comment]
Does something on Windows not allow you to `cp` the jar to a new file name?  Why was it ok with singlejar writing the new jar file?

If you want to use singlejar you'll have to enumerate all of the entries in the manifest and make `--deploy_manifest_line` entries for them.  That seemed worse than just using `jar` to update the manifest with the entry we need.

File: private/jvm_import.bzl [hunk 1]
```diff
@@ -13,20 +13,18 @@ def _jvm_import_impl(ctx):
...
+        inputs = [injar] + ctx.files._host_javabase,
         outputs = [outjar],
+        arguments = [],
+        command = " && ".join([
+            "cp {input_jar} {output_jar}".format(input_jar = injar.path, output_jar = outjar.path),
+            "echo 'Target-Label: {label}' > manifest.txt".format(label = ctx.label),
+            "{jar} ufm {output_jar} manifest.txt".format(
+                jar = "%s/bin/jar" % ctx.attr._host_javabase[java_common.JavaRuntimeInfo].java_home,
+                output_jar = outjar.path,
+            ),
```
Comment (jin) [inline private/jvm_import.bzl]: Prefer splitting this into two actions. One creates the manifest (with `ctx.action.expand_template`), and another that invokes `jar ufm` with the new manifest as an input.

**Comment by cheister** [comment]
@jin Is there anyway to enable `--subcommands=pretty_print` and get the test logs back from buildkite?  The Windows failures are pretty much a black box at this point.

**Comment by jin** [comment]
See https://github.com/bazelbuild/rules_jvm_external/pull/338/files

Got a little further with one last failing test: https://buildkite.com/bazel/rules-jvm-external-examples/builds/914#cc30e5b9-ec9f-4a14-b4e5-3f7c4ff7e5a6

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //src/test:main_activity_test
-----------------------------------------------------------------------------
JUnit4 Test Runner
.WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.robolectric.util.ReflectionHelpers$7 (file:/var/lib/buildkite-agent/.cache/bazel/_bazel_buildkite-agent/1f9c398f870b92e46230f621198a88ac/execroot/__main__/bazel-out/k8-fastbuild/bin/external/maven/v1/https/repo1.maven.org/maven2/org/robolectric/shadowapi/4.1/stamped_shadowapi-4.1.jar) to method java.lang.ClassLoader.getPackage(java.lang.String)
WARNING: Please consider reporting this to the maintainers of org.robolectric.util.ReflectionHelpers$7
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
E
Time: 0.22
There was 1 failure:
1) testOnCreateNotNull(com.example.bazel.MainActivityTest)
java.lang.SecurityException: Invalid signature file digest for Manifest main attributes
	at java.base/sun.security.util.SignatureFileVerifier.processImpl(SignatureFileVerifier.java:336)
	at java.base/sun.security.util.SignatureFileVerifier.process(SignatureFileVerifier.java:269)
	at java.base/java.util.jar.JarVerifier.processEntry(JarVerifier.java:316)
	at java.base/java.util.jar.JarVerifier.update(JarVerifier.java:230)
	at java.base/java.util.jar.JarFile.initializeVerifier(JarFile.java:757)
	at java.base/java.util.jar.JarFile.getInputStream(JarFile.java:838)
	at java.base/sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:167)
	at java.base/java.net.URLClassLoader.getResourceAsStream(URLClassLoader.java:328)
	at org.robolectric.internal.bytecode.SandboxClassLoader.getClassBytesAsStreamPreferringLocalUrls(SandboxClassLoader.java:104)
	at org.robolectric.internal.bytecode.SandboxClassLoader.getByteCode(SandboxClassLoader.java:161)
	at org.robolectric.internal.bytecode.SandboxClassLoader.maybeInstrumentClass(SandboxClassLoader.java:118)
	at org.robolectric.internal.bytecode.SandboxClassLoader.lambda$findClass$0(SandboxClassLoader.java:111)
	at org.robolectric.util.PerfStatsCollector.measure(PerfStatsCollector.java:50)
	at org.robolectric.internal.bytecode.SandboxClassLoader.findClass(SandboxClassLoader.java:110)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:3137)
	at java.base/java.lang.Class.getConstructor0(Class.java:3342)
	at java.base/java.lang.Class.getConstructor(Class.java:2151)
	at org.robolectric.RobolectricTestRunner.getHooksInterface(RobolectricTestRunner.java:553)
	at org.robolectric.RobolectricTestRunner.configureSandbox(RobolectricTestRunner.java:235)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:230)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
	at com.google.testing.junit.runner.internal.junit4.CancellableRequestFactory$CancellableRunner.run(CancellableRequestFactory.java:89)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at com.google.testing.junit.runner.junit4.JUnit4Runner.run(JUnit4Runner.java:112)
	at com.google.testing.junit.runner.BazelTestRunner.runTestsInSuite(BazelTestRunner.java:153)
	at com.google.testing.junit.runner.BazelTestRunner.main(BazelTestRunner.java:84)

FAILURES!!!
Tests run: 1,  Failures: 1


BazelTestRunner exiting with a return value of 1
JVM shutdown hooks (if any) will run now.
The JVM will exit once they complete.

-- JVM shutdown starting at 2019-12-19 19:40:19 --
```

**Comment by jin** [comment]
Ok, got all tests to pass in https://github.com/bazelbuild/rules_jvm_external/pull/338, with a really ugly hack to work around the invalid signature problem in the previous comment.

Feel free to patch that into this PR!

**Comment by cheister** [comment]
SG.  I'll pull your changes and squash everything.

**Comment by cheister** [comment]
Ok.  I pulled in your changes and found another problem with jars that have duplicate entries in them.  That causes the `jar umf` command to throw 
```
java.util.zip.ZipException: duplicate entry: META-INF/maven/
```
Apparently zip is ok with duplicate entries but jar is not.  I added a test jar for this in the PR.

I haven't had any luck finding an easy way to de-dup entries a jar file.  singlejar will remove duplicates but it will also overwrite the manfiest.mf file.    We could unzip the jar on the file system and re-jar it with the updated manifest.mf file?  That feels like more work than we should do but may work.  Another option is to write a java class that adds the manifest entry and use that to add better logic for catching exceptions?

For the jars with signature files it might be better to not update their manifest with Target-Label and leave them unstamped.  Seems like removing signatures might have unexpected consequences somewhere else.

**Comment by cheister** [comment]
ok.  I added a "best-effort" setup where the stamping would not fail if `jar umf` fails.  Currently for jars with duplicate entries you'll get output on the screen along the lines of 
```
INFO: From Stamping manifest of @regression_testing//:org_apache_beam_beam_sdks_java_core:
java.util.zip.ZipException: duplicate entry: META-INF/maven/
	at java.base/java.util.zip.ZipOutputStream.putNextEntry(ZipOutputStream.java:233)
	at java.base/java.util.jar.JarOutputStream.putNextEntry(JarOutputStream.java:109)
	at jdk.jartool/sun.tools.jar.Main.update(Main.java:972)
	at jdk.jartool/sun.tools.jar.Main.run(Main.java:366)
	at jdk.jartool/sun.tools.jar.Main.main(Main.java:1676)
```
But we could add `> /dev/null 2>&1` to the jar command and silence everything.  If we're ok with "best-effort" to stamp the jar this might be the way to go.  What do you think?

**Comment by jin** [comment]
> But we could add > /dev/null 2>&1 to the jar command and silence everything. If we're ok with "best-effort" to stamp the jar this might be the way to go. What do you think?

Sounds good, let's do this. 

> For the jars with signature files it might be better to not update their manifest with Target-Label and leave them unstamped. Seems like removing signatures might have unexpected consequences somewhere else.

This works as well, perhaps we can add a conditional check and only apply jar umf if signatures don't exist?

**Comment by cheister** [comment]
Ok, I added a test for signed jars and don't stamp the jar if it's signed.  Also squashed commits.

**Pull request merged**